base/inference.jl

```diff
@@ -1433,6 +1499,17 @@ function abstract_call(f::ANY, fargs, argtypes::Vector{Any}, vtypes::VarTable, s
 
     if istopfunction(tm, f, :promote_type) || istopfunction(tm, f, :typejoin)
         return Type
+    elseif length(argtypes) == 2 && istopfunction(tm, f, :typename)
```

Comment:
we should consider trying to handle this by marking the function `@pure` instead of doing a partial reimplementation here and trying to deal with all of the same edge cases